# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/svg/SVGIntegrationUtils.cpp
# Commit: beec388d4b0e
# Full Hash: beec388d4b0e6a50ac243ed86c0e3f1f48cbe744
# Author: Ashley Hale <ahale@mozilla.com>
# Date: 2024-06-25 09:36:02
# Regressor Bug: 1896503
# File Overlap Count: 1
# Description:
#   Bug 1896503 - Implement FilterInstance code to send SVG filter graph to WebRender r=mstange,gw
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D174208
# ==============================================================================

diff -r 2147e0051e23 -r beec388d4b0e layout/svg/SVGIntegrationUtils.cpp
--- a/layout/svg/SVGIntegrationUtils.cpp	Mon Jun 24 23:40:38 2024 +0000
+++ b/layout/svg/SVGIntegrationUtils.cpp	Mon Jun 24 23:45:05 2024 +0000
@@ -949,9 +949,17 @@
                                      opacity);
 }
 
-bool SVGIntegrationUtils::CreateWebRenderCSSFilters(
+WrFiltersStatus SVGIntegrationUtils::CreateWebRenderCSSFilters(
     Span<const StyleFilter> aFilters, nsIFrame* aFrame,
     WrFiltersHolder& aWrFilters) {
+  // Check if prefs are set to convert the CSS filters to SVG filters and use
+  // the new WebRender SVG filter rendering, rather than the existing CSS filter
+  // support
+  if (StaticPrefs::gfx_webrender_svg_filter_effects() &&
+      StaticPrefs::
+          gfx_webrender_svg_filter_effects_also_convert_css_filters()) {
+    return WrFiltersStatus::BLOB_FALLBACK;
+  }
   // All CSS filters are supported by WebRender. SVG filters are not fully
   // supported, those use NS_STYLE_FILTER_URL and are handled separately.
 
@@ -959,8 +967,10 @@
   // succeeded, and don't render any of them.
   if (aFilters.Length() >
       StaticPrefs::gfx_webrender_max_filter_ops_per_chain()) {
-    return true;
+    return WrFiltersStatus::DISABLED_FOR_PERFORMANCE;
   }
+  // Track status so we can do cleanup if unsupported filters are found.
+  WrFiltersStatus status = WrFiltersStatus::CHAIN;
   aWrFilters.filters.SetCapacity(aFilters.Length());
   auto& wrFilters = aWrFilters.filters;
   for (const StyleFilter& filter : aFilters) {
@@ -1024,28 +1034,39 @@
         break;
       }
       default:
-        return false;
+        status = WrFiltersStatus::BLOB_FALLBACK;
+        break;
+    }
+    if (status != WrFiltersStatus::CHAIN) {
+      break;
     }
   }
-
-  return true;
+  if (status != WrFiltersStatus::CHAIN) {
+    // Clean up the filters holder if we can't render filters this way.
+    aWrFilters = {};
+  }
+  return status;
 }
 
-bool SVGIntegrationUtils::BuildWebRenderFilters(
+WrFiltersStatus SVGIntegrationUtils::BuildWebRenderFilters(
     nsIFrame* aFilteredFrame, Span<const StyleFilter> aFilters,
     StyleFilterType aStyleFilterType, WrFiltersHolder& aWrFilters,
-    bool& aInitialized) {
-  return FilterInstance::BuildWebRenderFilters(
-      aFilteredFrame, aFilters, aStyleFilterType, aWrFilters, aInitialized);
+    const nsPoint& aOffsetForSVGFilters) {
+  return FilterInstance::BuildWebRenderFilters(aFilteredFrame, aFilters,
+                                               aStyleFilterType, aWrFilters,
+                                               aOffsetForSVGFilters);
 }
 
 bool SVGIntegrationUtils::CanCreateWebRenderFiltersForFrame(nsIFrame* aFrame) {
   WrFiltersHolder wrFilters;
   auto filterChain = aFrame->StyleEffects()->mFilters.AsSpan();
-  bool initialized = true;
-  return CreateWebRenderCSSFilters(filterChain, aFrame, wrFilters) ||
-         BuildWebRenderFilters(aFrame, filterChain, StyleFilterType::Filter,
-                               wrFilters, initialized);
+  WrFiltersStatus status =
+      CreateWebRenderCSSFilters(filterChain, aFrame, wrFilters);
+  if (status == WrFiltersStatus::BLOB_FALLBACK) {
+    status = BuildWebRenderFilters(aFrame, filterChain, StyleFilterType::Filter,
+                                   wrFilters, nsPoint());
+  }
+  return status == WrFiltersStatus::CHAIN || status == WrFiltersStatus::SVGFE;
 }
 
 bool SVGIntegrationUtils::UsesSVGEffectsNotSupportedInCompositor(